operator/pkg/ciliumendpointbatch/manager.go

```diff
@@ -0,0 +1,422 @@
+
+	for {
+		cebName = fmt.Sprintf("%s-%s-%s", cebNamePrefix, randomName(9), randomName(5))
+		if _, exists = cacheCeb[cebName]; !exists {
+			return cebName
+		}
+	}
+}
+
+// createCeb function create a new ceb and capacity to hold maximum ceps in a CEB.
+// This function called on 2 different scenarios.
+// 1) During runtime, when ceb manager decides to create a new ceb, it calls
+//    with an empty name. function generates random unique name and assign it to CEB
+// 2) During operator warm boot [after crash or software upgrade], batching manager
+//    create CEB, by passing unique name.
+func (c *cebManagerFcfs) createCeb(name string) *ceBatch {
+	var cebName string = name
+	if name == "" {
+		cebName = uniqueCeBatchName(c.cacheCeb)
+	}
+	ceb := &ceBatch{
+		ceb: &cilium_v2.CiliumEndpointBatch{
+			TypeMeta: meta_v1.TypeMeta{
+				Kind:       "CiliumEndpointBatch",
+				APIVersion: cilium_v2.SchemeGroupVersion.String(),
+			},
+			ObjectMeta: meta_v1.ObjectMeta{
+				Name: cebName,
+			},
+			Endpoints: make([]cilium_v2.CoreCiliumEndpoint, 0, c.maxCepsInCeb),
+		},
+		removedCeps: make(map[string]struct{}),
+	}
+	c.cacheCeb[cebName] = ceb
+	log.WithFields(logrus.Fields{
+		"ceb-name": cebName,
+	}).Debug("Generated ceb name")
+	return ceb
+}
+
+// If exists, remove Ceb object from cache. deleteCebFromCache called after successful removal from
+// apiserver.
+func (c *cebManagerFcfs) deleteCebFromCache(cebName string) {
+	if _, ok := c.cacheCeb[cebName]; !ok {
+		log.WithFields(logrus.Fields{
+			"ceb-name": cebName,
+		}).Debug("Failed to retrieve Ceb object in local cache.")
+		return
+	}
+	delete(c.cacheCeb, cebName)
+}
+
+// updateCebInCache function copies the ciliumEndpoint object in local cache. if isDeepCopy flag is set,
+// whole CoreCiliumEndpoint object stored in local cache.
+// There are two scenarios updateCebInCache is called.
+// 1) During operator warm boot[after crash or software upgrade], CEB controller sync CEB states from
+// api-server to cache. In this case, isDeepCopy set to true to copy entire CEP object locally.
+// 2) During runtime, reconciler sync current state with API server and update metadata only.
+// isDeepCopy flag is set to false.
+func (c *cebManagerFcfs) updateCebInCache(srcCeb *cilium_v2.CiliumEndpointBatch, isDeepCopy bool) {
+	if ceb, ok := c.cacheCeb[srcCeb.GetName()]; ok {
+		if !isDeepCopy {
+			ceb.ceb.ObjectMeta = srcCeb.ObjectMeta
+		} else {
+			srcCeb.DeepCopyInto(ceb.ceb)
+			for _, cep := range ceb.ceb.Endpoints {
+				// Update the cacheCepMap, to reflect all CEPs are packed in a CEB
+				c.cacheCepMap.insert(getCepNameFromCCEP(&cep), srcCeb.GetName())
+			}
+		}
+	}
+}
+
+// If available, getCebFromCache returns CiliumEndpointBatch object.
+func (c *cebManagerFcfs) getCebFromCache(cebName string) (*cilium_v2.CiliumEndpointBatch, error) {
+	if ceb, ok := c.cacheCeb[cebName]; ok {
+		outCeb := new(cilium_v2.CiliumEndpointBatch)
+		ceb.backendMutex.Lock()
+		ceb.ceb.DeepCopyInto(outCeb)
```

Comment:
I don't understand why we are deep copying this. The callers of this function don't modify the object so there's no need to deep copy it.

Reply:
CEB data in CEB Manager cache is the source of truth, also that is a moving variable. As CEP inserted/removed in a cluster, CEB data gets updated continuously. At some point of time, once we decided to sync the current CEB to api-server, we take the snapshot of current CEB and sync with api-server.
Before we reconcile a CEB, reconciler calls `getCebFromCache` to get `stable` CEB object for reconciling purpose.

Reply:
Why are we working with the `DeepCopy` functions if we can simply set the `ceb.ceb` whenver we want to set a new instance of the object? Can you point out a place where we modify the a `*cilium_v2.CiliumEndpointBatch` returned by this function? The only place I can find where the CEB is being modified is in the `updateCebInCache` function, for which we could simply set the entire pointer with the new struct received without having to do `DeepCopyInto`

Reply:
For reconciliation of a `CEB` we need the copy of the CEB. We can't send the original `ceb.ceb`, because that `CEB` being modified as `CEPs`  inserted and removed. 
If we use original `ceb.ceb` object for reconcile, we had noticed operator crash. I don't recall call stack.

Reply:
To avoid the cost of deep copy, made 2 different functions. One with `deepcopy` and other one returns `ceb.ceb`. This should solve reconciler requirement and avoid `deepcopies` in other scenarios.